### ISSUE TITLE
How to exclude and replace patches from installing

### ISSUE THREAD

**Issue by aangelinsf** [acquia/blt#1379]
My system information:
* Operating system type: macOS
* Operating system version: 10.12
* BLT version: 8.8.1

We are upgrading our code from 8.2.7 to 8.3.0 but it is failing. The new composer.json in Lightning wants:

            "drupal/core": {
                "2652138 - ImageFormatter does not support SVGs":
                "https://www.drupal.org/files/issues/2652138-47.patch"
            },

But that no longer passes on 8.3 and doesn't apply:
https://www.drupal.org/node/2652138

However #59 applies cleanly.

Also, the following:

  "1356276 - Allow profiles to provide a base/parent profile and load them in the correct order":
  "https://www.drupal.org/files/issues/1356276-302.patch"

doesn't apply and neither does #304.

It appears that the core code has moved on since this version of Lightning was cut so I would like to have composer skip it completely.

The closest I've been able to find in the composer documentation is perhaps to use the "provide" key then give it either the alternate patch (in the case of replacement) or just "*" (in the case of omitting it). This was a suggestion on a stackexchange question but in neither case have I been able to make this work.

How do I provide an alternate patch for composer or have it be omitted altogether. I am assuming that these need to be in the project composer.json because we don't have access to the Lightning composer.json during a deploy.

I saw that there is this task:
https://github.com/wikimedia/composer-merge-plugin/issues/142

perhaps that's relevant?

**Comment by grasmash** [comment]
The composer-patches package supports ignoring patches: https://github.com/cweagans/composer-patches#ignoring-patches

**Comment by grasmash** [comment]
Please re-open if this does not address your issue.

**Comment by aangelinsf** [comment]
I unsuccessfully tried ignoring patches per the instructions. Then I found this:
https://github.com/cweagans/composer-patches/issues/112

So others are finding that this feature doesn't work.

Someone has submitted a patch but it has not been accepted yet:
https://github.com/cweagans/composer-patches/pull/118

There may be some laborious ways around this (patching the code that patches?, create a custom version of composer-patches and host locally?).

**Comment by dmsmidt** [comment]
@aangelinsf, I've done what you wish to accomplish with the OpenSocial distribution. 

1. Use the patched composer-patches version you mentioned instead of the default.
Add the following to your root composer.json's repositories section:
`    {
      "type" : "vcs",
      "url" : "https://github.com/triquanta/composer-patches"
    },`
Then require the patched version like this: 
`    "cweagans/composer-patches": "dev-fix-ignore-dependency-patches as 1.6.1",`
2. Ignore the patch defined in the distro in the root composer.json (note ignore setup is slightly different in the new version)
3. Add a different patch in the root composer.json.

It would be great if you could test cweagans/composer-patches#118.
If you want to use it in production, I would advice to get the patched version from a repo you own.

**Comment by acontia** [comment]
Is [this](https://github.com/acquia/blt/issues/1379#issuecomment-295868485) still the only option to stop a patch from Lightning from being applied to Drupal core?

**Comment by mikemadison13** [comment]
what version of lightning are you trying to run @acontia ? part of the problem you may be running into is a mismatch in Drupal versions between what you are running and what Lightning expects

**Comment by acontia** [comment]
Yes that's the problem actually. I'm trying to run Drupal core 8.4.6 and Lightning 2.2.0.

Some of the patches provided by this version of Lightning don't apply to 8.4.6, but I can't upgrade to any newer version of Lightning (for now) because 2.2.1 requires migrating media entities to the core media module, which is giving me issues.

**Comment by mikemadison13** [comment]
I've (personally) had a TON of trouble getting dependencies a couple of layers down in composer to be ignored / changed at the top level. i'm not surprised that the ignore steps are giving you trouble.

have you tried working the media update process with the lightning team to unblock your upgrade path?

**Comment by acontia** [comment]
I am maintaining a multi-site setup and was trying to reduce the complexity of the upgrade by doing it in different phases. My idea was upgrading Core first and Lightning + Media in a next phase. Both tasks are complex enough that I preferred to do them separately.
I guess the only option at this point is either upgrading all together or raise my question in https://github.com/cweagans/composer-patches.
Thanks anyway for your help :)

**Comment by theneonlobster** [comment]
I'm trying to update the required version of Drupal Core to 8.5.x-dev and am running into this same blocker.

**Comment by hkorik** [comment]
@grasmash - I've been attempting to use https://github.com/cweagans/composer-patches#ignoring-patches to install Drupal Core 8.6.x-dev with BLT and ignore the following patch that is added with BLT 8.9.x and the patch refuses to be ignored no matter what I do.

```
      "drupal/core": {
        "Clear Twig caches on deploys": "https://www.drupal.org/files/issues/2752961-90.patch"
      }
```

There is an issue on [cweagans/composer-patches](https://github.com/cweagans/composer-patches/pull/210) but even with that commit merged into the version of cweagon I'm still running into the above issue.

Could it be that there is something else causing the core patch to consistently apply?

(FYI - Iv'e attempted the following with no luck:
* Added the patch to my projects composer file with ignoring it at the same time.
* Made sure tp clear composer cache prior to attempting to update core.)

Steps to reproduce:
- `composer create-project --no-interaction acquia/blt-project:8.x-dev [project]`
- Download patched version of cweagans/composer-patches https://github.com/cweagans/composer-patches/pull/210#issuecomment-387633773.
- Edit projects composer.json file with the below:
```
    "require": {
        "acquia/blt": "8.9.x-dev",
        "drupal/core": "8.6.x-dev",
        "cweagans/composer-patches": "dev-patch-1 as 1.6.5"
    }
```
```
    "extra": {
      "composer-exit-on-patch-failure": false,
      "enable-patching": true,
      "installer-types": [
        "bower-asset",
        "npm-asset"
    ],
      "installer-paths": {
        "docroot/core": [
            "type:drupal-core"
        ],
        "docroot/modules/contrib/{$name}": [
            "type:drupal-module"
        ],
        "docroot/modules/custom/{$name}": [
            "type:drupal-custom-module"
        ],
        "docroot/profiles/contrib/{$name}": [
            "type:drupal-profile"
        ],
        "docroot/profiles/custom/{$name}": [
            "type:drupal-custom-profile"
        ],
        "docroot/themes/contrib/{$name}": [
            "type:drupal-theme"
        ],
        "docroot/themes/custom/{$name}": [
            "type:drupal-custom-theme"
        ],
        "docroot/libraries/{$name}": [
            "type:drupal-library",
            "type:bower-asset",
            "type:npm-asset"
        ],
        "drush/contrib/{$name}": [
            "type:drupal-drush"
        ]
    },
      "merge-plugin": {
        "require": [
            "vendor/acquia/blt/composer.required.json"
        ],
        "merge-extra": true,
        "merge-extra-deep": true,
        "merge-scripts": true,
        "replace": false,
        "ignore-duplicates": true
    },
      "patches-ignore": {
          "drupal/core": {
              "Clear Twig caches on deploys": "https://www.drupal.org/files/issues/2752961-90.patch"
          }
      }
    }
```

- Attempt to run `composer update drupal/core`.

**Comment by hkorik** [comment]
@mikecrittenden - nope. 

I've been running into this issue on a project that's in development phase, so for now I've been manually removing the patch from the `vendor/acquia/blt/composer.required.json` file. (not the best approach I know..). 

Have you?

**Comment by mikecrittenden** [comment]
@hkorik so far I've been able to hold off on needing 8.6, but I will soon. This PR seems to work fine for me (https://github.com/acquia/blt/pull/2937) and I know that an 8.6 RC1 release is due this week and that 8.6 final is due September 5, so I assume BLT will be updating pretty soon.

**Comment by vuchkov** [comment]
https://github.com/acquia/blt/issues/1379#issuecomment-295868485 works perfectly for us. Thank you!
